CASSANDRA-20156: To improve accord interoperability test coverage, need to extend the harry model domain to handle more possible CQL states

dcapwell opened the pull request.

dcapwell committed:
Checkpointing a first draft of a AST Model

dcapwell commented on src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java:
```diff
@@ -201,8 +201,9 @@ protected ModificationStatement prepareInternal(ClientState state,
 
-                checkFalse(operations.appliesToRegularColumns(),
-                           "DELETE statements must restrict all PRIMARY KEY columns with equality relations in order to delete non static columns");
+                if (!operations.isEmpty())
```
going to fork this off into its own ticket... should be fixed outside of accord first...

dcapwell replied:
see https://issues.apache.org/jira/browse/CASSANDRA-20163

dcapwell replied:
Blake asked me to not do this for now, so ill update the model to document this rather than fix this code path

dcapwell commented on test/harry/main/org/apache/cassandra/harry/gen/ValueGenerators.java:
```diff
@@ -21,23 +21,29 @@
+                           Bijection<ClusteringKey> ckGen,
+                           Accessor<ClusteringKey> ckAccessor,
+
+                           List<? extends Bijection<? extends Object>> regularColumnGens,
+                           List<? extends Bijection<? extends Object>> staticColumnGens,
```
this is needed in order to allow the bijection cache to exist... working with `Object` is hard =D

dcapwell commented on test/harry/main/org/apache/cassandra/harry/gen/ValueGenerators.java:
```diff
@@ -54,22 +61,22 @@ public ValueGenerators(Bijections.Bijection<Object[]> pkGen,
     }
 
-    public Bijections.Bijection<Object> regularColumnGen(int idx)
+    public Bijection regularColumnGen(int idx)
```
`?` and `Object` don't work together.. and `Object` requires casting to the constructor... since the actual usage of this is the untyped case, I moved to untyped and avoid generics...

dcapwell commented on test/harry/main/org/apache/cassandra/harry/model/PartitionState.java:
```diff
@@ -223,7 +232,7 @@ private RowState updateRowState(RowState currentState, IntFunction<Bijections.Bi
                 assert lts >= currentState.lts[i] : String.format("Out-of-order LTS: %d. Max seen: %s", lts, currentState.lts[i]); // sanity check; we're iterating in lts order
 
-                if (currentState.lts[i] == lts)
+                if (lts != MagicConstants.NO_TIMESTAMP && currentState.lts[i] == lts)
```
right now my test doesn't own the timestamps, so we disable.  When you disable then we end up trying to compare objects rather than replacing with the latest... needed this step to avoid such a problem

dcapwell commented on test/unit/org/apache/cassandra/cql3/DeleteTest.java:
```diff
@@ -46,7 +46,7 @@ public void prepare() throws Exception
 
         session.execute("drop keyspace if exists junit;");
-        session.execute("create keyspace junit WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 2 };");
+        session.execute("create keyspace junit WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };");
```
paxos isn't happy with rf2 as we can't reach quorum...

dcapwell commented on test/unit/org/apache/cassandra/cql3/DeleteTest.java:
```diff
@@ -93,6 +93,33 @@ public void prepare() throws Exception
     }
 
+    @Test
+    public void deleteCasWithStatic()
```
ill pull this out into its own ticket... this should be fixed in mainline branches and not accord

aratno commented on test/harry/main/org/apache/cassandra/harry/model/BytesPartitionState.java:
```diff
@@ -0,0 +1,555 @@
+        private final Factory factory;
+        @Nullable
+        public final Clustering<ByteBuffer> key;
+        public final Token token;
```
Why not DecoratedKey to handle token collisions?

dcapwell replied:
`BufferDecoratedKey` is `(ByteBuffer key, Token token)`, yet this class is `(Clustering<ByteBuffer> key, Token token)`, this change is mostly to avoid having to split the key.

lets say the table is `pk0, pk1, pk2` and you do a search for `pk0 = 42`, you then need to split the key (if using `DecoratedKey`) to get access to `pk0`, but if you use `Clustering` then you are working with `ByteBuffer[]` so can just access the value from the array.

dcapwell commented on src/java/org/apache/cassandra/db/filter/RowFilter.java:
```diff
@@ -224,7 +224,7 @@ protected BaseRowIterator<?> applyToPartition(BaseRowIterator<?> partition)
                 // Short-circuit all partitions that won't match based on static and partition keys
                 for (Expression e : partitionLevelExpressions)
-                    if (!e.isSatisfiedBy(metadata, partition.partitionKey(), partition.staticRow()))
+                    if (!e.isSatisfiedBy(metadata, partition.partitionKey(), partition.staticRow(), nowInSec))
```
this is https://issues.apache.org/jira/browse/CASSANDRA-20108, needed to make the tests stable

dcapwell commented on test/unit/org/apache/cassandra/cql3/ast/AssignmentOperator.java:
```diff
@@ -109,4 +109,14 @@ public AbstractType<?> type()
     }
+
+    @Override
+    public Expression visit(Visitor v)
```
this should be in https://issues.apache.org/jira/browse/CASSANDRA-20155

dcapwell commented on test/unit/org/apache/cassandra/cql3/ast/CQLFormatter.java:
```diff
@@ -0,0 +1,128 @@
+
+import static org.apache.cassandra.cql3.ast.Elements.newLine;
+
+public interface CQLFormatter
```
this should be in https://issues.apache.org/jira/browse/CASSANDRA-20155

dcapwell commented:
the diff should be smaller, its pulling in some `ast` changes that should be in https://issues.apache.org/jira/browse/CASSANDRA-20155 instead

dcapwell commented on test/unit/org/apache/cassandra/utils/GeneratorsTest.java:
```diff
@@ -45,4 +55,42 @@ public void dnsDomainName()
     }
+
+    @Test
+    public void asciiDeterministic()
```
I added these tests as I was finding test failures and when I copied the text I would get different results.  It turns out that IntelliJ wasn't properly handling displaying the text properly but it pushed me down a rabbit hole trying to figure out if deterministic generation wasn't working...

dcapwell committed:
fixed backport

dcapwell committed:
fixed backport

dcapwell committed:
checkpoint but need to give up, java is fighting mixed case too much

dcapwell committed:
drop the inet test as it cant really do what is needed

dcapwell committed:
Refactor the ast package to enable harry model based testing

patch by David Capwell; reviewed by Abe Ratnofsky, Caleb Rackliffe for CASSANDRA-20155

dcapwell commented:
now the parent branch is merged, closing this in favor of https://github.com/apache/cassandra/pull/3785
